### PR TITLE
[None][feat] Support named logits post-processors in the Triton LLMAPI backend

### LINKED ISSUE #6202: Support `LogitsPostProcessorConfig` in Triton Backend
https://github.com/NVIDIA/TensorRT-LLM/blob/98428f330e2f1d1b5606ca55ec4d30f0970dcab4/triton_backend/inflight_batcher_llm/src/model_instance_state.cc#L697

`LogitsPostProcessorConfig` is currently hardcoded to null. Please add support for logits so that we users can set it in config.pbtxt

If a maintainer can give pointers, I can implement this change and raise a PR.

Thank you.

### PR DESCRIPTION
**Pull request by harshal-96**
## Description

Closes #6202.

Requests to the llmapi Triton model cannot use logits post-processing (for example token biasing), because there is no way to route a processor into `SamplingParams` from a Triton request. As noted in the issue thread, this makes features like token biasing impossible through the Triton backend even though the LLM API itself supports `SamplingParams(logits_processor=...)`.

This PR adds an opt-in, name-based mechanism following the direction sketched in the issue (configuration through the LLMAPI model's `model.yaml`), with one deliberate difference: processors are referenced by python import spec (`module.path:attribute`) and resolved through the regular import machinery, instead of eval'ing lambda strings from the config. Only code already installed on the Triton server can be referenced.

- `model.yaml`: optional `logits_post_processors` section mapping names to import specs. Specs resolve to a `tensorrt_llm.sampling_params.LogitsProcessor` instance or class; classes are instantiated once at model load. The section is consumed by the backend and excluded from the LLM engine args.
- `config.pbtxt`: new optional string input `sampling_param_logits_post_processor_name`. A request selects a configured processor by name; it is passed to `SamplingParams` as `logits_processor`. Unknown names fail with a `TritonModelException` listing the configured processors.
- Behavior is fully opt-in: with no `logits_post_processors` section and no request input, nothing changes.

## Test Coverage

- New CPU unit tests in `tests/unittest/triton_backend/test_llmapi_helpers.py` (11 cases): registry loading for classes, instances, and nested attributes; every error path (non-mapping section, malformed spec, missing module, missing attribute, non-callable target); request-side resolution including absent input, empty name, unknown name, and lookup. The tests stub `triton_python_backend_utils`, so they run without a Triton server or GPU.
- End-to-end validation through a running Triton server is not included; happy to extend the L0_backend_trtllm coverage if maintainers point me at the preferred place.

## Impact

No change for existing deployments (new config section and request input are both optional). The per-request cost when unused is one absent-tensor lookup.

## PR Checklist

- [x] PR title follows the `[JIRA/NVBUG/None][type]` format
- [x] DCO sign-off on the commit
- [x] Unit tests added
- [ ] CI to be triggered by maintainers (`/bot run`)

## Summary

Adds opt-in, name-based logits post-processor support to the Triton LLMAPI backend.

- Loads processors from `model.yaml` using `module:attribute` import specifications.
- Supports processor instances and classes. Instantiates classes at model load.
- Validates processor construction and picklability.
- Adds the optional `sampling_param_logits_post_processor_name` request input.
- Applies the selected processor through `SamplingParams`.
- Reports configured processor names for unknown selections.
- Preserves existing behavior when the feature is unused.
- Resolves issue `#6202`.

## Dev Engineer Review

- `helpers.py` validates import specifications, processor types, empty components, construction failures, and picklability.
- Import and attribute lookup errors preserve their original causes.
- Request lookup handles absent, empty, and unknown processor names.
- `model.py` excludes `logits_post_processors` from engine arguments.
- `model.yaml` documents shared-instance, statelessness, serialization, and lifetime requirements.
- `config.pbtxt` defines the optional request input.
- The implementation uses Python import machinery and avoids evaluating configuration strings.
- The test-list entry uses the correct test path and format.

## QA Engineer Review

Added or modified test functions cover:

- Processor loading from classes and instances.
- Nested attribute lookup.
- Invalid import specifications.
- Constructor failures.
- Unpicklable instances.
- Non-`ImportError` module-load failures.
- Request-side processor lookup.
- Model YAML configuration plumbing.

The tests are registered in `tests/integration/test_lists/test-db/l0_a10.yml` for CI.

**Verdict: sufficient.**

### CODE REVIEW

File: tests/unittest/triton_backend/test_llmapi_helpers.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (brnguyen2) [inline tests/unittest/triton_backend/test_llmapi_helpers.py:1]: This is a new test directory, and nothing under `tests/integration/test_lists/test-db/` references `triton_backend`, so these tests will not run in pre-merge CI. They're CPU-only, so add `unittest/triton_backend/test_llmapi_helpers.py` to a CPU test list (e.g. `l0_a10.yml`) so regressions in `helpers.py` are actually caught.

File: triton_backend/all_models/llmapi/tensorrt_llm/1/helpers.py [hunk 2]
```diff
@@ -177,6 +183,95 @@ def get_lora_request_from_request(request, batch_size=1, batch_index=0):
...
+                f"logits_post_processors['{name}']: cannot import module "
+                f"'{module_name}': {e}") from e
+        for attr in attr_path.split("."):
+            try:
+                obj = getattr(obj, attr)
+            except AttributeError as e:
+                raise pb_utils.TritonModelException(
+                    f"logits_post_processors['{name}']: module "
+                    f"'{module_name}' has no attribute '{attr_path}'") from e
+        if isinstance(obj, type):
```
Comment (brnguyen2) [inline triton_backend/all_models/llmapi/tensorrt_llm/1/helpers.py:238]: Two failure modes escape as raw tracebacks during `initialize()` rather than as a `TritonModelException`:

- `obj()` here raises `TypeError` if the class needs constructor arguments — a likely user mistake, since a token-biasing processor usually wants config.
- `importlib.import_module` at line 222 only catches `ImportError`; a module that raises anything else at import time (e.g. missing GPU, config error) propagates unwrapped.

Both surface to the operator as an opaque model-load failure. Wrapping them the same way as the other paths, with the processor name in the message, makes the misconfiguration obvious.

File: triton_backend/all_models/llmapi/tensorrt_llm/1/model.yaml [hunk 3]
```diff
@@ -8,6 +8,18 @@ backend: "pytorch"
 tensor_parallel_size: 1
 pipeline_parallel_size: 1
 
+# ======= Logits Post-Processors (optional) =======
+# Named logits post-processors that requests can select via the optional
+# 'sampling_param_logits_post_processor_name' input tensor. Each entry maps a
+# name to a python import spec 'module.path:attribute' resolving to a
+# tensorrt_llm.sampling_params.LogitsProcessor instance or class (classes are
+# instantiated once, with no arguments, at model load). The referenced module
```
Comment (brnguyen2) [inline triton_backend/all_models/llmapi/tensorrt_llm/1/model.yaml:16]: Worth stating the lifetime and serialization contract here, since the current wording ("instantiated once at model load") suggests a long-lived shared object with persistent state, which is not what happens.

The single instance is shared by all concurrent requests, and depending on executor mode it is pickled into the worker process per request along with `SamplingParams` (`tensorrt_llm/executor/ipc.py`). So: the processor must be picklable; state mutated inside `__call__` does not persist across requests or propagate back; and a processor carrying large state (a full-vocab bias tensor, say) pays that serialization cost on every request. A sentence covering "stateless across requests, picklable, keep the payload small" would save users a confusing debugging session.

**Review by brnguyen2 — COMMENTED** [review]
The design choice — import specs resolved through `importlib` instead of eval'ing config strings — is the right call, and the opt-in path is clean. Two things to settle before merge.

**The new unit test file isn't wired into CI.** `tests/unittest/triton_backend/test_llmapi_helpers.py` is a new directory and no `tests/integration/test_lists/test-db/*.yml` references `triton_backend`, so the 11 tests never execute in pre-merge. Since they are CPU-only, add `unittest/triton_backend/test_llmapi_helpers.py` to a CPU list (e.g. `l0_a10.yml`); otherwise the file is dead weight the first time someone changes `helpers.py`.

**Document the processor lifetime contract.** One instance is created at model load and shared by every request. Depending on the executor mode, the `SamplingParams` (and with it the processor object) is pickled to the worker process per request over the ZMQ IPC path (`tensorrt_llm/executor/ipc.py`), so: the processor must be picklable, per-request mutations made in the worker do not persist back or across requests, and a large processor (e.g. a full-vocab bias table) pays that serialization cost on every request. None of that is visible from the `model.yaml` comment as written; a couple of lines there ("processors must be stateless across requests and picklable") would keep users out of a confusing failure mode. A `pickle.dumps` smoke check at load time would also turn a per-request runtime error into a clear startup error, if you think it's worth the code.

The description matches the diff, and the end-to-end gap is honestly stated. A minimal L0_backend_trtllm case exercising one configured processor plus one unknown-name rejection would be the natural follow-up, but I wouldn't block on it given the unit coverage.

**Comment by harshal-96** [comment]
Thanks for the thorough review @brnguyen2, all three points addressed in 8f015a7:

1. **CI wiring:** `unittest/triton_backend/test_llmapi_helpers.py` is now registered in the `l0_a10` CPU pre-merge list.
2. **Error wrapping:** zero-argument instantiation failures and non-ImportError import-time exceptions both now raise `TritonModelException` naming the processor, with regression tests for each.
3. **Lifetime contract:** documented in `model.yaml` (single shared instance, stateless across requests, picklable, large payloads pay per-request serialization). I also took your optional suggestion: `pickle.dumps` runs once at model load, so an unpicklable processor fails at startup with a clear message instead of on the first request. Test added.

18 unit tests total, all passing. The offer stands on the `L0_backend_trtllm` integration case (one configured processor plus one unknown-name rejection); happy to add it in this PR or as a follow-up, whichever you prefer.

**Review by StanleySun639 — APPROVED** [review]

**Review by tburt-nv — APPROVED** [review]
